### ISSUE TITLE
one-dimensional coupling flows do not work

### ISSUE THREAD

**Issue by jackerschott** [VincentStimper/normalizing-flows#58]
For the following example:
```
import numpy as np
import torch as T

import normflows as nf

base_dist = nf.distributions.base.DiagGaussian(1, trainable=False)

flows = []
for _ in range(16):
    flows += [nf.flows.CoupledRationalQuadraticSpline(
            num_input_channels=1, num_context_channels=1,
            num_blocks=5, num_hidden_channels=256, reverse_mask=True)]
    flows += [nf.flows.Permute(1)]

network = nf.ConditionalNormalizingFlow(q0=base_dist, flows=flows)
network.train()
network = network.to('cuda')

optim = T.optim.Adam(network.parameters())
for i in range(2):
    x = T.randn(10_000, 1).to('cuda')
    c = T.randn(10_000, 1).to('cuda')

    loss = network.forward_kld(x, c)
    print('loss:', loss)

    optim.zero_grad()
    loss.backward()
    optim.step()

    print(any([T.any(T.isnan(p)) for p in network.parameters()]))
```
I encounter the following issues:
1. I get a NaN-loss and NaN network parameters after the first optimizer step.
2. When I remove reverse_mask=True, I get a shape error.
3. I get a shape error with or without reverse_mask=True when I use no context dimension.

I specifically only get 1 when I perform the computation on the GPU, not on the CPU. Presumably because of precision issues?
It is clear why 2 happens, since in that case our one-dimension gets mapped to the identity split, instead of the transform split.

What I would expect instead is that the one-dimensional case works and works out of the box, specifically without having to set reverse_mask=True.

To be fair, one could easily argue that one shouldn't use coupling flows for 1d and instead use an autoregressive flow, which doesn't really change anything (except that it works without problems now). Following this thought, I would simply suggest to throw an error if one attempts to initialize a coupling flow with only one channel .

**Comment by sipie800** [comment]
I reproduce your result. And your set is not of 1d but a scalar actually.
Increasing the size can resolve this, though I'm not sure if it's proper to use a scalar in this pipeline.

**Comment by VincentStimper** [comment]
Hi @jackerschott and @sipie800,

Coupling flows are inherently designed to operate on at least two dimensions, since in a coupling layer the dimensions are split into two disjoint sets, i.e. one that is transformed and another that is kept constant but used for conditioning, see e.g. the [Real NVP paper](https://arxiv.org/abs/1605.08803). Hence, you cannot apply a coupling flow to a one dimensional target distribution. Instead, you could use an autoregressive [affine](https://github.com/VincentStimper/normalizing-flows/blob/c609d119bb7a547231e5cce09008422454fbb125/normflows/flows/affine/autoregressive.py#L50) or [spline](https://github.com/VincentStimper/normalizing-flows/blob/c609d119bb7a547231e5cce09008422454fbb125/normflows/flows/neural_spline/wrapper.py#L186) flow, which will use constant parameters in the one dimensional case. Moreover, using a [residual flow](https://github.com/VincentStimper/normalizing-flows/blob/c609d119bb7a547231e5cce09008422454fbb125/normflows/flows/residual.py#L12) is possible, and this would be perhaps the most expressive flow for a one dimensional problem.

Best regards,
Vincent
